### PR TITLE
Core and heatmap

### PR DESCRIPTION
**Pull request by cryogenian**
This is the very basic thing to leaflet.js 

I made this because updating existing lib from 0.7 -> 0.11 and generalizing it has almost same complexety. 

Also ported Leaflet/Leaflet.heat with funny `Ref` emitting code. 

Depends on garyb/purescript-dom-classy#6

@natefaubion Please take a look.

### CODE REVIEW

File: src/Leaflet/Core/Plugin.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+exports.onAddRemove_ = function(Nothing, Just, onAdd, onRemove, layer, map) {
+    return function() {
+        var res = {value: Nothing};
```
Comment (natefaubion) [inline src/Leaflet/Core/Plugin.js]: We shouldn't be making assumptions about the representation of refs. We should just do this stuff in PS land.

```purescript
onAddRemove init finish l lf = liftEff do
  ref <- newRef Nothing
  let
    onAddFn = mkFn2 \a b -> do
      val <- init a b
      writeRef ref val
    onRemoveFn = ...
  runFn6 onAddRemove_ ..
  pure ref
```

File: src/Leaflet/Core/Layer.purs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  ⇒ f T.LatLng
+  → Record r1
+  → m T.Polygon
+polygon lls r =
+  liftEff $ runFn3 polygon_ (foldMap (A.singleton ∘ converter.convertLatLng) lls) converter r
+
+rectangle
+  ∷ ∀ e r1 r2 m
+  . MonadEff (dom ∷ DOM|e) m
+  ⇒ Union r1 r2 (T.PolylineConf ())
```
Comment (natefaubion) [inline src/Leaflet/Core/Layer.purs:209]: So on stuff like this, there are no _required_ options for the confs, right? For any of them `{}` is a valid configuration?

**Review by natefaubion — CHANGES_REQUESTED** [review]

File: src/Leaflet/Core/Layer.purs (hunk 2 above)
Comment (cryogenian) [reply src/Leaflet/Core/Layer.purs:209]: Yep, empty option map is valid

File: src/Leaflet/Core/Plugin.js (hunk 1 above)
Comment (cryogenian) [reply src/Leaflet/Core/Plugin.js]: :+1:

**Comment by cryogenian** [comment]
I've abstracted `Ref`, so, js doesn't know how refs work. Although creation and writing to it are located in js code. 
@natefaubion Is this ok? I can create `Ref` in PS and leave only modifying in JS or just remove all this ref stuff at all until we actually need it

**Comment by natefaubion** [comment]
The main thing is just if there is a _need_ for it to be in JS. We shouldn't put anything in JS that can be easily done in PS. At a glance, it looks like there is no reason that `addRemove_` can't just take regular callbacks, and we can do all the `Ref` coordination in PS. Is that the case, or does the FFI rely on the `Ref` business?

**Comment by cryogenian** [comment]
As far as I understand leaflet code, `onAdd` _should_ (contract or whatever) return `this`, so modification of of ref should be done in js. Although instead of `Nothing` and `newRef` I can put there ref with `Nothing` :D

**Comment by cryogenian** [comment]
Ah! Yep, `init` must be called in JS handler because leaflet/layer state could change

**Comment by cryogenian** [comment]
I think this will work

**Review by natefaubion — APPROVED** [review]

**Pull request merged**